### DISCUSSION TITLE
How to assign different preset to different directory in flat config?

### DISCUSSION THREAD

**Discussion by foray1010** [eslint/eslint#16472] [Q&A]
Hi, I have been designing my eslint config to adapt different use cases, for example, an eslint config `A` for node.js script and an eslint config `B` for react only code

By using the legacy config approach, I can easily assign different config by different directory, but in flat config, as there is only single config file, I can still manage to config different rules per directory by following code:

```js
export default [
  { files: ["scripts/**"], rules: {...} },
  { files: ["src/**"], rules: {...} },
]
```

The issue raise when I have to use different preset per directory, because the preset outputs an array of flat config now, and nested config is not supported by flat config. Is there any way I could achieve this?

example preset: `A`
```js
export default [...someRules]
```

example preset: `B`
```js
export default [...completelyDifferentSetOfRulesAndLanguageOptionsAndEvenParser]
```

In the `eslint.config.js`

```js
import A from './A.js'
import B from './A.js'

export default [
  { files: ["scripts/**"], extends: A }, // <- currently no way to do it
  { files: ["src/**"], extends: B }, // <- currently no way to do it
]
```

**Comment by nzakas** [comment]
You're correct, there is no way to do this, as we don't know what "extending" an array of configs actually means. That said, there are plenty of ways to mix and match flat configs. Can you share what shareable config you're trying to use? That would be helpful in thinking through the right way to use it.

**Reply by foray1010** [reply]
Thanks for replying, maybe I should explain my real world use case:

## Creating shareable config

1. I have a shareable config under package name [@foray1010/eslint-config](https://github.com/foray1010/common-presets/tree/master/packages/eslint-config#installation), there are two config now (legacy format): `@foray1010/eslint-config` for node.js code and `@foray1010/eslint-config/react` for react-related browser code
2. I am migrating this shareable config to flat config - If I get it right, I should export an array (flat config) in each config and import it in other project

   example:
   
   ```js
   const nodeConfig = [
     "eslint:recommended",
     ...someNodeJsSpecificRules,
   ]
   export default nodeConfig
   ```

## Consuming the sharable config

1. Before flat config, I consume the config in the frontend project in the following way:
    1. I put an `.eslintrc.cjs` in root and point to `@foray1010/eslint-config`
    2. I also put an `.eslintrc.cjs` in `/src` directory and point to `@foray1010/eslint-config/react`
2. Now when I migrate to flat config, I notice I can only have one `eslint.config.js` in the root, but I cannot reproduce the previous behaviour because I can't specifiy different sharable config in different directory:

    ```js
    import nodeConfig from '@foray1010/eslint-config'
    import reactConfig from '@foray1010/eslint-config/react'

    export default [
      { files: ["/**"], ignores: ["src/**"] }, // <- want to use nodeConfig
      { files: ["src/**"] }, // <- want to use reactConfig
    ]
    ```

**Reply by nzakas** [reply]
So here is what I would recommend:

```js
import nodeConfig from '@foray1010/eslint-config'
import reactConfig from '@foray1010/eslint-config/react'

function applyConfig(filesConfig, sharedConfig) {
  return sharedConfig
    .filter(config => typeof config !== "string")
    .map(config => ({
      ...config,
      ...filesConfig
    }));
}

export default [
   "eslint:recommended",  // <- apply to all

   applyConfig({
     files: ["/**"],
     ignores: ["src/**"]  
   }, nodeConfig),

   applyConfig({
     files: ["src/**"]
   }, reactConfig),
]
```

Part of the benefit of flat config is that because configs are just arrays of objects, you can manipulate them however you want.

**Reply by foray1010** [reply]
Thanks @nzakas, I thought of this method before but later I realize it may cause some issues, for example:

1. We may specify `files` in the shared config itself, e.g. only apply some rules for TypeScript, the `applyConfig` will overwrite the `files` and apply the rules to all files instead.
2. It will interfere other config, e.g.:

   ```js
   // @foray1010/eslint-config
   export default [
     ...someRules,
     {
         ignores: ["**/*.test.js"],
     },
     ...someRules,
   ]
   ```
   
   Then all rules in `@foray1010/eslint-config/react` will not be executed in test files, which is unexpected
  
Of coz, we can somehow merge the `files`/`ignores` globs in `applyConfig`, it could solve `1` but not `2`. IMO it is quite error-prone to let developer to implement their own solution. (I presume I am not the only one separating eslint config per use case, I see similar pattern in [eslint-config-xo](https://github.com/xojs/eslint-config-xo) and [eslint-config-xo-react](https://github.com/xojs/eslint-config-xo-react))

**Reply by nzakas** [reply]
A similar problem exists in the eslintrc format, where `overrides` with `extends` will merge the configs together including `files` and `ignores`, such that the result is difficult to understand. 

The situation with flat config is better because you have complete control over your configs and how you want to merge them. If there's an error, you can fix them. With the `extends` model, you have no control and need to accept however ESLint decides to merge them which, again, is complex for `files` and `ignores`.

For flat config, it's recommended that shared configs avoid `files` and `ignores` unless absolutely necessary. For instance, if you want to share a group of rules, then do that without `files` and `ignores`. If you want to share a config in your company or across multiple projects, then it's okay to include `files` and `ignores`, just make sure your documentation specifies that.

**Comment by nzakas** [comment]
Related: https://github.com/eslint/eslint/issues/16537
